ast/desugar/Desugar.cc

```diff
@@ -1885,6 +1920,44 @@ TreePtr node2TreeImpl(DesugarContext dctx, unique_ptr<parser::Node> what) {
                 ast::cast_tree_nonnull<ast::Send>(res).flags.isPrivateOk = true;
                 result = std::move(res);
             },
+            [&](parser::CaseMatch *caseMatch) {
+                // Create a local var to store the expression used in each match clause
+                auto exprLoc = caseMatch->expr->loc;
+                auto exprName = dctx.ctx.state.freshNameUnique(core::UniqueNameKind::Desugar, core::Names::assignTemp(),
+                                                               dctx.uniqueCounter++);
+                auto exprVar = MK::Assign(exprLoc, exprName, node2TreeImpl(dctx, std::move(caseMatch->expr)));
+
+                // Desugar the `else` block
+                TreePtr res = node2TreeImpl(dctx, std::move(caseMatch->elseBody));
+
+                // Desugar each `in` as an `if` branch calling `Magic.<pattern-match>()`
+                for (auto it = caseMatch->inBodies.rbegin(); it != caseMatch->inBodies.rend(); ++it) {
+                    auto inPattern = parser::cast_node<parser::InPattern>(it->get());
+                    ENFORCE(inPattern != nullptr, "case pattern without a in?");
+
+                    // Keep the `in` body for the `then` body of the new `if`
+                    auto pattern = std::move(inPattern->pattern);
+                    auto body = node2TreeImpl(dctx, std::move(inPattern->body));
+
+                    // Desugar match variables found inside the pattern
+                    InsSeq::STATS_store vars;
+                    desugarPatternMatchingVars(vars, dctx, pattern);
+                    if (!vars.empty()) {
+                        body = MK::InsSeq(pattern->loc, std::move(vars), std::move(body));
+                    }
+
+                    // Call `Magic.<pattern-match>(expr, TODO pattern, TODO guard)`
```

Comment:
This comment mentions three arguments, but I only see one argument passed to `<Magic>.<pattern-match>`?

Reply:
Also ditto above—can we use `ast::MK::Unimplemented` for the pattern match expr if we're not going to implement it with an intrinsic or something else?

Reply:
Replaced by `RaiseUnimplemented`.